### PR TITLE
feat(Scripts): Implement generic script loader and registry macros

### PR DESCRIPTION
**Pull request by Kitzunu**
* cherry-pick commit (https://github.com/TrinityCore/TrinityCore/commit/fb87ac8e8da3d215d4988b8c2fa25d7a98cf9ee0)
* cherry-pick commit (https://github.com/TrinityCore/TrinityCore/commit/7d4fbf706ae19481468901ccd65c497fcabe56bf)
* cherry-pick commit (https://github.com/TrinityCore/TrinityCore/commit/75a6a7a0ad48874e75dd71f766dad3707341bb32)

Co-Authored-By: Treeston <14020072+Treeston@users.noreply.github.com>
Co-Authored-By: Shauren <shauren.trinity@gmail.com>

## Changes Proposed:
-  Implement generic script loader and registry macros
- Implement 
-  Turn Marrowgar into new system

## Tests Performed:

- Builds
- Tested in-game

## How to Test the Changes:

1. Test that Marrowgar works (AKA his script, so talk and spells)

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Review by Nyeriah — APPROVED** [review]

File: src/server/game/Scripting/ScriptMgr.h [hunk 1]
```diff
@@ -1947,6 +1949,85 @@ class ScriptMgr
...
+private:
+    SpellScript* GetSpellScript() const override
+    {
+        if constexpr (!std::is_same_v<SpellScriptType, Acore::find_type_end>)
+        {
+            return Acore::new_from_tuple<SpellScriptType>(_args);
+        }
+        else
+        {
+            return nullptr;
```
Comment (Kitzunu) [inline src/server/game/Scripting/ScriptMgr.h:1980]:
```suggestion
            return nullptr
```

File: src/server/game/Scripting/ScriptMgr.h [hunk 2]
```diff
@@ -1947,6 +1949,85 @@ class ScriptMgr
...
+
+    AuraScript* GetAuraScript() const override
+    {
+        if constexpr (!std::is_same_v<AuraScriptType, Acore::find_type_end>)
+        {
+            return Acore::new_from_tuple<AuraScriptType>(_args);
+        }
+        else
+        {
+            return nullptr
```
Comment (Kitzunu) [inline src/server/game/Scripting/ScriptMgr.h:1992]:
```suggestion
            return nullptr;
```

**Comment by Winfidonarleyan** [comment]
Very good

**Comment by Kitzunu** [comment]
@azerothcore/testers please

**Comment by Nyeriah** [comment]
Tested, works.
Skärmbild 2021-11-07 011148

**Comment by Kitzunu** [comment]
Thank you 💚

**Pull request merged**